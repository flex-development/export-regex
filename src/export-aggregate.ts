/**
 * @file EXPORT_AGGREGATE_REGEX
 * @module export-regex/aggregate
 */

/**
 * Aggregate `export` statement regex. Ignores matches in comments.
 *
 * @see https://regex101.com/r/JtvRUt
 * @see https://developer.mozilla.org/docs/Web/JavaScript/Reference/Statements/export#re-exporting_aggregating
 *
 * @example
 *  import { EXPORT_AGGREGATE_REGEX } from '@flex-development/export-regex'
 *  import { dedent } from 'ts-dedent'
 *
 *  const code: string = dedent`
 *    export { defineBuildConfig, type BuildConfig } from "#src"
 *    export type {
 *      JsonObject,
 *      LiteralUnion,
 *      Nullable
 *    } from '@flex-development/tutils'
 *    export * as constants from "./constants"
 *    export * from './interfaces'
 *  `
 *
 *  const print = (matches: IterableIterator<RegExpMatchArray>): void => {
 *    console.debug([...matches].map(match => omit(match, ['input'])))
 *  }
 *
 *  print(code.matchAll(EXPORT_AGGREGATE_REGEX))
 *  // [
 *  //   {
 *  //     '0': 'export { defineBuildConfig, type BuildConfig } from "#src"',
 *  //     '1': undefined,
 *  //     '2': '{ defineBuildConfig, type BuildConfig }',
 *  //     '3': '#src',
 *  //     index: 0,
 *  //     groups: [Object: null prototype] {
 *  //       type: undefined,
 *  //       exports: '{ defineBuildConfig, type BuildConfig }',
 *  //       specifier: '#src'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': 'export * as constants from "./constants"',
 *  //     '1': undefined,
 *  //     '2': '* as constants',
 *  //     '3': './constants',
 *  //     index: 149,
 *  //     groups: [Object: null prototype] {
 *  //       type: undefined,
 *  //       exports: '* as constants',
 *  //       specifier: './constants'
 *  //     }
 *  //   },
 *  //   {
 *  //     '0': "export * from './interfaces'",
 *  //     '1': undefined,
 *  //     '2': '*',
 *  //     '3': './interfaces',
 *  //     index: 190,
 *  //     groups: [Object: null prototype] {
 *  //       type: undefined,
 *  //       exports: '*',
 *  //       specifier: './interfaces'
 *  //     }
 *  //   }
 *  // ]
 *
 * @const {RegExp} EXPORT_AGGREGATE_REGEX
 */
const EXPORT_AGGREGATE_REGEX: RegExp =
  /(?<=^|[\n;](?:[\t ]*(?:\w+ )?)?)export(?:(?:\s+(?<type>type)\s*)|\s*)(?<exports>(?:\*(?:\s+as\s+\S+)?)|\S+|(?:{[\w\t\n\r "$'*,./{}-]+?}))\s*from\s*["']\s*(?<specifier>(?:(?<='\s*)[^']*[^\s'](?=\s*'))|(?:(?<="\s*)[^"]*[^\s"](?=\s*")))\s*["']/g

export default EXPORT_AGGREGATE_REGEX
