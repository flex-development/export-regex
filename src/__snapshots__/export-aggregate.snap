// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match named export(s) in multi-line statement 1`] = `
{
  "0": "export {
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
} from './lib'",
  "1": undefined,
  "2": "{
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
}",
  "3": "./lib",
  "groups": {
    "exports": "{
  addFive,
  addFour,
  addThree,
  addTwo,
  squareFive,
  squareFour,
  squareThree,
  squareTwo
}",
    "specifier": "./lib",
    "type": undefined,
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match named export(s) in single-line statement 1`] = `
{
  "0": "export { defineBuildConfig, type BuildConfig } from \\"#src\\"",
  "1": undefined,
  "2": "{ defineBuildConfig, type BuildConfig }",
  "3": "#src",
  "groups": {
    "exports": "{ defineBuildConfig, type BuildConfig }",
    "specifier": "#src",
    "type": undefined,
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match named type export(s) in multi-line statement 1`] = `
{
  "0": "export type {
  JsonObject,
  LiteralUnion,
  Nullable
} from '@flex-development/tutils'",
  "1": "type",
  "2": "{
  JsonObject,
  LiteralUnion,
  Nullable
}",
  "3": "@flex-development/tutils",
  "groups": {
    "exports": "{
  JsonObject,
  LiteralUnion,
  Nullable
}",
    "specifier": "@flex-development/tutils",
    "type": "type",
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match named type export(s) in single-line statement 1`] = `
{
  "0": "export type { default as Options } from \\"./options\\"",
  "1": "type",
  "2": "{ default as Options }",
  "3": "./options",
  "groups": {
    "exports": "{ default as Options }",
    "specifier": "./options",
    "type": "type",
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match namespace export 1`] = `
{
  "0": "export * from './interfaces'",
  "1": undefined,
  "2": "*",
  "3": "./interfaces",
  "groups": {
    "exports": "*",
    "specifier": "./interfaces",
    "type": undefined,
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match namespace export with alias 1`] = `
{
  "0": "export * as constants from \\"./constants\\"",
  "1": undefined,
  "2": "* as constants",
  "3": "./constants",
  "groups": {
    "exports": "* as constants",
    "specifier": "./constants",
    "type": undefined,
  },
}
`;

exports[`unit:EXPORT_AGGREGATE_REGEX > exports > should match nested export statement(s) in declaration file 1`] = `
[
  {
    "0": "export { default } from './make'",
    "1": undefined,
    "2": "{ default }",
    "3": "./make",
    "groups": {
      "exports": "{ default }",
      "specifier": "./make",
      "type": undefined,
    },
  },
  {
    "0": "export {
    DEFAULTS,
    plugin as default,
    type Options
  } from './plugin'",
    "1": undefined,
    "2": "{
    DEFAULTS,
    plugin as default,
    type Options
  }",
    "3": "./plugin",
    "groups": {
      "exports": "{
    DEFAULTS,
    plugin as default,
    type Options
  }",
      "specifier": "./plugin",
      "type": undefined,
    },
  },
  {
    "0": "export type { Config, Result } from './interfaces'",
    "1": "type",
    "2": "{ Config, Result }",
    "3": "./interfaces",
    "groups": {
      "exports": "{ Config, Result }",
      "specifier": "./interfaces",
      "type": "type",
    },
  },
  {
    "0": "export type { default as Options } from './options'",
    "1": "type",
    "2": "{ default as Options }",
    "3": "./options",
    "groups": {
      "exports": "{ default as Options }",
      "specifier": "./options",
      "type": "type",
    },
  },
  {
    "0": "export type Foo from '#foo'",
    "1": "type",
    "2": "Foo",
    "3": "#foo",
    "groups": {
      "exports": "Foo",
      "specifier": "#foo",
      "type": "type",
    },
  },
]
`;
